### PR TITLE
Densenet generator/discriminator

### PR DESCRIPTION
**Pull request by martyn**
Adds support for a custom densenet d/g.  Other tweaks and fixes snuck in too.

### CODE REVIEW

File: hypergan/__init__.py [hunk 1]
```diff
@@ -59,19 +60,22 @@
 hc.set('dtype', tf.float32) #The data type to use in our GAN.  Only float32 is supported at the moment
 
 # Generator configuration
-hc.set("generator.z", 100) # the size of the encoding.  Encoder is set by the 'encoder' property, but could just be a random_uniform
```
Comment (martyn) [inline hypergan/__init__.py]: Don't change these defaults.

File: hypergan/discriminators/pyramid_nostride_discriminator.py [hunk 2]
```diff
@@ -19,7 +19,7 @@ def discriminator(config, x, g, xs, gs):
           net = batch_norm(config['batch_size']*2, name='d_expand_bn_'+str(i))(net)
       net = activation(net)
       # APPEND xs[i] and gs[i]
-      if(i < len(xs) and i > 0):
+      if(i < len(xs) and i > 1):
```
Comment (martyn) [inline hypergan/discriminators/pyramid_nostride_discriminator.py]: Don't change this in this branch.

File: hypergan/util/hc_tf.py [hunk 3]
```diff
@@ -293,18 +293,25 @@ def block_deconv(result, activation, batch_size,id,name, output_channels=None, s
...
     if(dropout):
+        print("DOROPOUT")
         z = get_tensor('original_z')
         mask = linear(z, s[1]*s[2]*s[3], scope="g_lin_proj_mask")
         mask = tf.reshape(mask, result.get_shape())
         result *= tf.nn.sigmoid(mask)
+        set_tensor('z_proj_tanh', result)
+
+    if(resize):
+        result = tf.image.resize_images(result, resize, 1)
```
Comment (martyn) [inline hypergan/util/hc_tf.py]: Move to --resize

File: hypergan/__init__.py [hunk 4]
```diff
@@ -136,7 +144,7 @@
 hc.set("examples_per_epoch", 30000/4)
 
 #TODO category/bernouilli
-categories = [[]]#[[2]+[2]+build_categories_config(30)]
+categories = [[2,2,2,2,3,34,4,4,4,4,4,4,4,10,10,10,10,10,10,10,10,26,26,26,26,26,26,26,10,10,10,10,2,2,2,2,2,3,3,3,3,3,3,3,3,26,40,26]]#[[2]+[2]+build_categories_config(30)]
```
Comment (mikkel) [inline hypergan/__init__.py]: do we want this?

File: hypergan/generators/resize_conv.py [hunk 5]
```diff
@@ -21,22 +21,23 @@ def generator(config, net):
     net = block_conv(net, tf.nn.tanh, batch_size, 'identity', 'g_layers_init', output_channels=int(net.get_shape()[3]), filter=3, dropout=get_tensor('dropout'))
     z_proj = net
 
+    z_proj = None
```
Comment (mikkel) [inline hypergan/generators/resize_conv.py]: dont change this file

File: hypergan/loaders/image_loader.py [hunk 6]
```diff
@@ -45,12 +46,12 @@ def labelled_image_tensors_from_directory(directory, batch_size, channels=3, for
...
-  tf.Tensor.set_shape(reshaped_image, [None, None, None])
+  #reshaped_image = tf.identity(img)
+  #tf.Tensor.set_shape(reshaped_image, [None, None, None])
 
-  reshaped_image = hypergan.loaders.resize_image_patch.resize_image_with_crop_or_pad(reshaped_image,
-                                                         224, 224, dynamic_shape=True)
-  reshaped_image = tf.reshape(reshaped_image, [1,224,224,channels])
+  #reshaped_image = hypergan.loaders.resize_image_patch.resize_image_with_crop_or_pad(reshaped_image,
+  #                                                       224, 224, dynamic_shape=True)
+  #reshaped_image = tf.reshape(reshaped_image, [1,224,224,channels])
```
Comment (mikkel) [inline hypergan/loaders/image_loader.py]: what does this remove?

File: hypergan/util/gan_server.py [hunk 7]
```diff
@@ -41,6 +41,25 @@ def sample_batch(self, sample_file):
        
        #plot(self.config, sample, sample_file)
 
+    def sample_zeros(self, sample_file):
```
Comment (mikkel) [inline hypergan/util/gan_server.py]: move this into a sampler?

File: hypergan/util/hc_tf.py [hunk 8]
```diff
@@ -293,18 +293,25 @@ def block_deconv(result, activation, batch_size,id,name, output_channels=None, s
...
     size = int(result.get_shape()[-1])
-    s = result.get_shape()
+    result = activation(result)
     if(batch_norm is not None):
         result = batch_norm(batch_size, name=name+'bn')(result)
-    print("DROPOUT IS", dropout)
-    result = activation(result)
+    s = result.get_shape()
     if(dropout):
+        print("DOROPOUT")
```
Comment (mikkel) [inline hypergan/util/hc_tf.py]: reeomove

File: hypergan/util/hc_tf.py [hunk 9]
```diff
@@ -293,18 +293,25 @@ def block_deconv(result, activation, batch_size,id,name, output_channels=None, s
...
 
-def block_conv(result, activation, batch_size,id,name, output_channels=None, stride=2, noise_shape=None, dtype=tf.float32,filter=3, batch_norm=None, dropout=None):
+def block_conv(result, activation, batch_size,id,name, resize=None, output_channels=None, stride=2, noise_shape=None, dtype=tf.float32,filter=3, batch_norm=None, dropout=None, reshaped_z_proj=None):
     size = int(result.get_shape()[-1])
-    s = result.get_shape()
+    result = activation(result)
     if(batch_norm is not None):
         result = batch_norm(batch_size, name=name+'bn')(result)
-    print("DROPOUT IS", dropout)
-    result = activation(result)
```
Comment (mikkel) [inline hypergan/util/hc_tf.py]: does reversing these work well?

File: hypergan/util/hc_tf.py (hunk 8 above)
Comment (mikkel) [inline hypergan/util/hc_tf.py]: oookkk

**Pull request merged**